mixnet/node/src/lib.rs

```diff
@@ -77,205 +75,241 @@ impl MixnetNode {
     ) -> Result<()> {
         tracing::debug!("Delaying the packet for {delay:?}");
         tokio::time::sleep(delay.to_duration()).await;
 
-        Self::forward(
-            pool,
-            max_retries,
-            retry_delay,
-            Body::new_sphinx(packet),
-            NymNodeRoutingAddress::try_from(next_node_addr)?,
-        )
-        .await
+        self.forward(packet, NymNodeRoutingAddress::try_from(next_node_addr)?)
+            .await
     }
 
     async fn forward_payload_to_destination(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        payload: Payload,
+        &self,
+        payload: Body,
         destination_addr: DestinationAddressBytes,
     ) -> Result<()> {
         tracing::debug!("Forwarding final payload to destination mixnode");
 
-        Self::forward(
-            pool,
-            max_retries,
-            retry_delay,
-            Body::new_final_payload(payload),
+        self.forward(
+            payload,
             NymNodeRoutingAddress::try_from_bytes(&destination_addr.as_bytes())?,
         )
         .await
     }
 
-    async fn forward(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        body: Body,
-        to: NymNodeRoutingAddress,
-    ) -> Result<()> {
+    async fn forward(&self, pkt: Body, to: NymNodeRoutingAddress) -> Result<()> {
         let addr = SocketAddr::from(to);
-        let arc_socket = pool.get_or_init(&addr).await?;
-
-        if let Err(e) = {
-            let mut socket = arc_socket.lock().await;
-            body.write(&mut *socket).await
-        } {
-            tracing::error!("Failed to forward packet to {addr} with error: {e}. Retrying...");
-            return mixnet_protocol::retry_backoff(
-                addr,
-                max_retries,
-                retry_delay,
-                body,
-                arc_socket,
-            )
+
+        self.message_tx.send(Packet::new(addr, pkt))?;
+        Ok(())
+    }
+}
+
+struct PacketForwarder {
+    config: MixnetNodeConfig,
+    message_rx: mpsc::UnboundedReceiver<Packet>,
+    message_tx: mpsc::UnboundedSender<Packet>,
+    connections: HashMap<SocketAddr, TcpStream>,
+}
+
+impl PacketForwarder {
+    pub fn new(
+        message_tx: mpsc::UnboundedSender<Packet>,
+        message_rx: mpsc::UnboundedReceiver<Packet>,
+        config: MixnetNodeConfig,
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                pkt = self.message_rx.recv() => {
+                    if let Some(pkt) = pkt {
+                        self.send(pkt).await;
+                    } else {
+                        // Channel closed, we should shutdown the packet forwarder thread
```

Comment:
nit: not sure it's possible since `PacketForwarder` internally stores the sending half